### ISSUE TITLE
merge_to over successive cycles causes reproducible precision loss

### ISSUE THREAD

**Issue by bghira** [KohakuBlueleaf/LyCORIS#264]
hello again,

when we're running multi-merge inference, the results are stellar and the performance is much improved over that of `apply_to`, but a problem persists where the model has to be completely reloaded to effectively and safely change the adapter stack.

this is because using `merge_to(-1)` will cause a small but perceptible drift in output upon the first revert of the model weights back to the baseline state. from here, each invocation of `merge_to()` and `merge_to(-1)` cause further destruction to the base model weights.

here is a MRE:

```py
"""Reproduce precision drift from repeated merge/unmerge operations."""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from itertools import chain
from pathlib import Path
import sys
from typing import Iterable

import torch
import torch.nn as nn

if __package__ is None or __package__ == "":  # pragma: no cover
    sys.path.append(str(Path(__file__).resolve().parents[1]))

from lycoris import LycorisNetwork, create_lycoris


@dataclass
class DriftResult:
    cycle: int
    max_abs: float
    mean_abs: float


def build_wrappers(
    module: nn.Module,
    algos: Iterable[str],
    *,
    rank: int,
    alpha: float,
    seed: int,
) -> list[LycorisNetwork]:
    """Create and attach wrappers with randomised weights."""

    torch.manual_seed(seed)
    wrappers: list[LycorisNetwork] = []
    for algo in algos:
        wrapper = create_lycoris(
            module,
            multiplier=1.0,
            linear_dim=rank,
            linear_alpha=alpha,
            algo=algo,
        )
        wrapper.apply_to()

        # Populate weights with random values so merge_to has observable effect.
        with torch.no_grad():
            for param in wrapper.parameters():
                torch.nn.init.normal_(param)

        wrappers.append(wrapper)
    return wrappers


def run_cycles(
    base: nn.Linear,
    wrappers: Iterable[LycorisNetwork],
    *,
    cycles: int,
    report_every: int,
    merge_value: float,
) -> list[DriftResult]:
    """Perform repeated merge/unmerge cycles and record weight drift."""

    wrappers = list(wrappers)
    original_weight = base.weight.detach().clone()
    original_bias = base.bias.detach().clone() if base.bias is not None else None

    results: list[DriftResult] = []

    def measure() -> DriftResult:
        delta = base.weight - original_weight
        max_abs = delta.abs().max().item()
        mean_abs = delta.abs().mean().item()
        return DriftResult(step + 1, max_abs, mean_abs)

    for step in range(cycles):
        for wrapper in wrappers:
            wrapper.merge_to(merge_value)
        for wrapper in reversed(wrappers):
            wrapper.merge_to(-merge_value)

        if (step + 1) % report_every == 0 or step == cycles - 1:
            results.append(measure())

    # Sanity check bias as well if present
    if base.bias is not None and original_bias is not None:
        bias_delta = (base.bias - original_bias).abs().max().item()
        if bias_delta != 0.0:
            print(f"Warning: bias drift detected ({bias_delta:g})")

    return results


def format_summary(results: Iterable[DriftResult]) -> str:
    lines = ["cycle,max_abs,mean_abs"]
    for item in results:
        lines.append(f"{item.cycle},{item.max_abs:.12e},{item.mean_abs:.12e}")
    return "\n".join(lines)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Demonstrate weight drift from repeated merge_to/merge_to(-1) calls",
    )
    parser.add_argument(
        "--algos",
        nargs="+",
        default=["lora", "loha"],
        help="Sequence of algorithms to stack on the same module",
    )
    parser.add_argument("--cycles", type=int, default=1000, help="Number of merge cycles")
    parser.add_argument(
        "--report-every",
        type=int,
        default=100,
        help="Interval (in cycles) at which to report drift",
    )
    parser.add_argument("--seed", type=int, default=0, help="Torch RNG seed")
    parser.add_argument("--rank", type=int, default=4, help="Adapter rank")
    parser.add_argument("--alpha", type=float, default=1.0, help="Adapter alpha")
    parser.add_argument(
        "--merge-value",
        type=float,
        default=1.0,
        help="Positive weight passed to merge_to (inverse uses the negative)",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    base = nn.Linear(32, 32)
    wrappers = build_wrappers(
        base,
        args.algos,
        rank=args.rank,
        alpha=args.alpha,
        seed=args.seed,
    )

    results = run_cycles(
        base,
        wrappers,
        cycles=args.cycles,
        report_every=args.report_every,
        merge_value=args.merge_value,
    )

    print(format_summary(results))


if __name__ == "__main__":
    main()
```

we can run the example like:

```bash
python tools/merge_precision_drift.py --cycles 200 --report-every 50
```

giving us the results:

```csv
cycle,max_abs,mean_abs
50,2.682209014893e-07,1.274148075936e-08
100,2.682209014893e-07,1.274148075936e-08
150,2.682209014893e-07,1.274148075936e-08
200,2.682209014893e-07,1.274148075936e-08
```

**Comment by KohakuBlueleaf** [comment]
resolved as #267  merged
